### PR TITLE
Filtered Event Streams

### PR DESCRIPTION
**Pull request by mbolt35**
* Allows `EventStream[T]` to be conditionally selected at dispatch time. This allows better listening capabilities per event type. 
  * Added `NewFilteredEventStream(condition EventCondition[T]) EventStream[T]` to dispatcher.
  * Added `AddFilteredEventHandler(handler EventHandler[T], condition EventCondition[T]) HandlerID` to dispatcher
* Made race fixes for some of the unit tests

### CODE REVIEW

File: set.go [hunk 1]
```diff
@@ -133,6 +136,25 @@ func (s *lockingSet[T]) ToSlice() []T {
...
+	if l == 0 {
+		return nil
+	}
+
+	var slice []T
+	for k := range s.m {
+		if predicate(k) {
+			slice = append(slice, k)
+		}
+	}
```
Comment (mbolt35) [inline set.go:154]: So, I didn't put much additional thought into this, but it did occur to me that we could always create a slice with capacity of `len(s.m)` which would decrease the total number of reallocs that could occur on append, but we'd almost always over-allocate if there were actual filtered events. Given that this number is small most of the time, I don't really think it's a huge problem, but worth a passing thought if you have one :)

File: dispatcher.go [hunk 2]
```diff
@@ -120,11 +157,20 @@ type multicastDispatcher[T any] struct {
 // AddEventHandler adds a new event handler method that is called whenever an event T is dispatched. A
 // unique HandlerID is returned that can be used to remove the handler.
 func (md *multicastDispatcher[T]) AddEventHandler(handler EventHandler[T]) HandlerID {
+	return md.AddFilteredEventHandler(handler, nil)
+}
+
+// AddFilteredEventHandler adds a new event handler method that is called whenever a dispatched event T
+// passes the provided condition. A unique HandlerID is returned that can be used to remove the handler.
+func (md *multicastDispatcher[T]) AddFilteredEventHandler(handler EventHandler[T], condition EventCondition[T]) HandlerID {
 	handlerID := HandlerID(uuid.NewString())
```
Comment (mbolt35) [inline dispatcher.go:166]: I'm thinking that UUID is probably too much here -- thoughts? I feel like we could get away with something like ETL has, a 5-char random generation.

File: dispatcher.go [hunk 3]
```diff
@@ -214,7 +267,7 @@ func (md *multicastDispatcher[T]) CloseEventStreams() {
 
 // newMulticastDispatcher creates a new Dispatcher[T]
 func newMulticastDispatcher[T any]() Dispatcher[T] {
-	in := make(chan *dispatchedEvent[T], 5)
+	in := make(chan *dispatchedEvent[T], 20)
```
Comment (mbolt35) [inline dispatcher.go:270]: Increasing this, while dispatching 5 events of the same type at once seemed semi-aggressive, I'd rather provide more buffer room here. I want to make sure we maintain a good dispatch to receiver ratio. I will formalize a benchmark at some point.

File: dispatcher.go (hunk 2 above)
Comment (nikovacevic) [reply dispatcher.go:166]: Proooobably? But I don't think UUID is too much overhead, and it makes me feel fuzzy and safe. I dig it.

Comment (michaelmdresser) [reply dispatcher.go:166]: I'm with Niko!

**Review by michaelmdresser — APPROVED** [review]
I dig it! Expanded README with examples is awesome.

File: dispatcher.go [hunk 4]
```diff
@@ -34,38 +45,55 @@ type Dispatcher[T any] interface {
 	// Dispatch broadcasts the T event to any subscribed listeners asynchronously.
 	Dispatch(event T)
 
-	// DispatchSync is a special dispatch scenario which will block if any
-	// listeners are not yet receiving events. This should be used if you
-	// need to guarantee that all receivers are processing events before
-	// continuing.
+	// DispatchSync is a special dispatch scneario which will block if any listeners are not yet receiving
```
Comment (nikovacevic) [inline dispatcher.go]: scneario --> scenario

**Review by nikovacevic — APPROVED** [review]
LGTM!

**Comment by Adam-Stack-PM** [comment]
@nikovacevic Is this PR targeting v1.94? If so, can I add the label?

**Pull request merged**